ls1intum/Artemis
# Enhance programming exercise details page with lifecycle timeline (readonly) and repo + build plans

### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de and locally.

### Motivation and Context

This change enhances the details page with two useful elements that are currently only present in the update view. For the reason of consistent information presentation the lifecycle timeline as well as the preview of the build + repo preview should be integrated in the detail page as well. 
This PR tackles issue #2621.

### Description
I refactored the two existing components for programming exercises.
**`ProgrammingExerciseLifecycleComponent`**: was refactored having a dedicated module in the shared folder of programming-exercises to be used in the `ExerciseDetailsModule` that is then integrated in both pages, the update and the details page of a programming exercise. To disallow for changes in the details page in contrast to the update page, a `readOnly` flag is passed. It manipulates the buttons on the lifecycle timeline in a way to not be clickable/hoverable, not show the date-picker and not remove the buttons.

**`ProgrammingExercisePlansAndRepositoriesPreviewComponent`**: stays in the `ProgrammingExerciseUpdateModule` as this module was imported in the `ProgrammingExerciseManagementModule` anyways. It is therefore possible to add the component in the template of the `ProgrammingExerciseDetailComponent` without further changes.

### Steps for Testing


1. Log in to Artemis as instructor
2. Navigate to Course Management
3. Choose a course with programming-exercise (or create a new programming-exercise if not present) and navigate to the detail page of the programming-exercise (`course-management/11/programming-exercises/1111`)
4. Check if the lifecycle timeline is present in a readOnly mode, i.e. nothing can be clicked and changed
5. Check if the information shown in the lifecycle (correct dates if set, orange button if not set) is correct
6. Check if the build and repo preview is present
7. Check if the information shown in the build and repo preview (i.e. names) is correct

In contrast to checking the newly introduced elements in the details page, it should be assured, that the use of these components in the update page are not affected negatively.

1. Log in to Artemis as instructor
2. Navigate to Course Management
3. Choose a course
4. Create a new programming exercise (`course-management/11/programming-exercises/new`)
5. Check if the lifecycle timeline is present and can be used interactively, i.e. new dates can be set, set dates can be changed and steps can be removed
6. Check if the build and repo preview is present

### Test Coverage
No files to be tested touched... or tell me if I miss something.

### Screenshots
Integration of Lifecycle in details page:
![image](https://user-images.githubusercontent.com/41366522/115201573-c3f55900-a0f5-11eb-82b1-0e3844191351.png)

Integration of Repo and Build preview in details plage:
![image](https://user-images.githubusercontent.com/41366522/115201673-e2f3eb00-a0f5-11eb-80a6-59e861e8748b.png)
